File: Server/src/main/java/org/example/services/JwtService.java
package org.example.services;

import io.jsonwebtoken.*;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import org.example.entities.user.UserEntity;
import org.example.entities.user.UserRoleEntity;
import org.example.repositories.UserRoleRepo;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.security.Key;
import java.util.Date;
import java.util.List;

@Service
public class JwtService {
    private final UserRoleRepo userRoleRepo;

    // A key to encrypt a token
    private static final String jwtSecretString = "404E635266556A586E3272357538782F413F4428472B4B6250645367566B5970";
    private static final byte[] jwtSecret = Decoders.BASE64.decode(jwtSecretString);

    // Owner of the token. Write your own domain later
    private static final String jwtIssuer = "step.io";

    @Autowired
    public JwtService(UserRoleRepo userRoleRepo) {
        this.userRoleRepo = userRoleRepo;
    }

    // Method to generate an access token for a specific user
    public String generateAccessToken(UserEntity user) {
        List<UserRoleEntity> roles = userRoleRepo.findByUser(user);
        return Jwts.builder()
                .setSubject(user.getId() + "," + user.getUsername())
                .claim("email", user.getUsername())
                .claim("roles",
                        roles.stream()
                                .map(role -> role.getRole().getName())
                                .toArray(String[]::new)
                )
                .setIssuer(jwtIssuer)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + 7 * 24 * 60 * 60 * 1000))
                .signWith(signInKey(), SignatureAlgorithm.HS256)
                .compact();
    }

    private Key signInKey() {
        return Keys.hmacShaKeyFor(jwtSecret);
    }

    // Get the user ID from the token
    public String getUserId(String token) {
        Claims claims = Jwts.parserBuilder()
                .setSigningKey(jwtSecret)
                .build()
                .parseClaimsJws(token)
                .getBody();
        return claims.getSubject().split(",")[0];
    }

    // Get the username from the token
    public String getUsername(String token) {
        Claims claims = Jwts.parserBuilder()
                .setSigningKey(jwtSecret)
                .build()
                .parseClaimsJws(token)
                .getBody();
        return claims.getSubject().split(",")[1];
    }

    // Get the expiration date of the token
    public Date getExpirationDate(String token) {
        Claims claims = Jwts.parserBuilder()
                .setSigningKey(jwtSecret)
                .build()
                .parseClaimsJws(token)
                .getBody();
        return claims.getExpiration();
    }

    // Validate whether our token is valid and was issued by our server
    public boolean validate(String token) {
        try {
            Jwts.parserBuilder().setSigningKey(jwtSecret).build().parseClaimsJws(token);
            return true;
        } catch (MalformedJwtException ex) {
            System.out.println("Invalid JWT token - " + ex.getMessage());
        } catch (ExpiredJwtException ex) {
            System.out.println("Expired JWT token - " + ex.getMessage());
        } catch (UnsupportedJwtException ex) {
            System.out.println("Unsupported JWT token - " + ex.getMessage());
        } catch (IllegalArgumentException ex) {
            System.out.println("JWT claims string is empty - " + ex.getMessage());
        } catch (Exception ex) {
            System.out.println("Invalid JWT signature - " + ex.getMessage());
        }
        return false;
    }
}

File: Server/src/main/java/org/example/services/JwtService.kt
package org.example.services

import io.jsonwebtoken.*
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import org.example.entities.UserEntity
import org.example.entities.UserRoleEntity
import org.example.repositories.UserRoleRepo
import org.springframework.stereotype.Service
import java.security.Key
import java.util.*

@Service
class JwtService(
    private val userRoleRepo: UserRoleRepo
) {
    // ключ, яким ми шифруємо (будь-які букви чи цифри)
    private val jwtSecret =
        "404E635266556A586E3272357538782F413F4428472B4B6250645367566B5970"
    //вказує хто власник цього токена. Можна вписати ім'я свого домена
    private val jwtIssuer = "step.io"

    //метод призначений для того, щоб для визначеного юзера зробити jwt token
    fun generateAccessToken(user: UserEntity): String {
        val roles: List<UserRoleEntity> = userRoleRepo.findByUser(user)
        return Jwts.builder()
            .setSubject(String.format("%s,%s", user.id, user.username))
            .claim("email", user.username) //.claim("image", user.getImage())
            .claim("roles",
                roles//витягується списочок ролей, які є у юзера
                    .map { role: UserRoleEntity -> role.role.name }
                    .toTypedArray()
            )
            .setIssuer(jwtIssuer) //записуємо хто власник токена
            .setIssuedAt(Date(System.currentTimeMillis())) //коли був створений токен
            .setExpiration(Date(System.currentTimeMillis() + 7 * 24 * 60 * 60 * 1000)) // 1 week  зазначаємо скільки часу буде жити токен
            .signWith(signInKey, SignatureAlgorithm.HS256) //шифруємо токен за допомогою сікретного ключа
            .compact()
    }

    private val signInKey: Key
        get() {
            val keyBytes: ByteArray = Decoders.BASE64.decode(jwtSecret)
            return Keys.hmacShaKeyFor(keyBytes)
        }

    // з токена можна витягнути Id юзера
    fun getUserId(token: String?): String {
        val claims: Claims = Jwts.parser()
            .setSigningKey(jwtSecret) // перевіряється чи цей токен видавався нашим серваком
            .parseClaimsJws(token)
            .body
        return claims.subject.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            .get(0) //з токена бере перший елемент Id
    }

    // з токена можна витягнути username юзера
    fun getUsername(token: String?): String {
        val claims: Claims = Jwts.parser()
            .setSigningKey(jwtSecret)
            .parseClaimsJws(token)
            .body
        return claims.subject.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray().get(1)
    }

    // метод повертає дату до якої живе токен
    fun getExpirationDate(token: String?): Date {
        val claims: Claims = Jwts.parser()
            .setSigningKey(jwtSecret)
            .parseClaimsJws(token)
            .body
        return claims.expiration
    }

    //перевіряє чи наш токен валідний і чи видавався нашим сервером
    fun validate(token: String?): Boolean {
        try {
            Jwts.parser().setSigningKey(jwtSecret).parseClaimsJws(token)
            return true
        } catch (ex: SignatureException) {
            println("Invalid JWT signature - " + ex.message)
        } catch (ex: MalformedJwtException) {
            println("Invalid JWT token - " + ex.message)
        } catch (ex: ExpiredJwtException) {
            println("Expired JWT token - " + ex.message)
        } catch (ex: UnsupportedJwtException) {
            println("Unsupported JWT token - " + ex.message)
        } catch (ex: IllegalArgumentException) {
            println("JWT claims string is empty - " + ex.message)
        }
        return false
    }
}
